core/arch/arch.c

```diff
@@ -3633,8 +3633,35 @@ priv_mcontext_to_dr_mcontext(dr_mcontext_t *dst, priv_mcontext_t *src)
                     return false;
                 memcpy(&dst->opmask, &src->opmask, sizeof(dst->opmask));
             }
+#elif defined(AARCHXX)
+#    ifdef X64
+            /* The first and so far (August 2023) only DR_MC_MULTIMEDIA related
+             * change to AARCH64's machine context was addition of the Scalable
+             * Vector Extension (SVE) vector registers (DR_REG_Z0->DR_REG_Z31),
+             * predicate registers (DR_REG_P0->DR_REG_P15) and first fault
+             * register (DR_REG_FFR). Further SVE2 support did not change
+             * these. To avoid breaking backward compatibility with pre-SVE
+             * clients, we only copy the first 128 bits of each vector, i.e.
+             * the v8 NEON SIMD registers DR_REG_Q0->DR_REG_Q31.
+             */
+            if (sizeof(dst->simd) < sizeof(src->simd)) {
+                /* For pre-SVE and SVE builds:
+                 * MCXT_NUM_SIMD_SLOTS = MCXT_NUM_SIMD_SVE_SLOTS = 32.
+                 * with the only difference in simd[] being the size of each
+                 * simd slot.
+                 */
+                for (int i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; ++i)
+                    memcpy(&dst->simd[i], &src->simd[i], 16);
```

Comment:
This is also not going to do the right thing: a user with compiled code that uses the old definition will have an array of smaller simd elements.  Probably what we want is a union definition kept around for the old layout and then we'll have to compute the array offsets manually; or possibly keeping around the entire old dr_mcontext_t.  This extra complexity is because the dr_simd_t type was directly changed.  Generally we're only appending new fields, and then the old fields are identical and we just do a size check as above to see whether a new field exists.  Changing the size of elements in the array adds the complexity.

Reply:
As discussed offline, since we planned to bump the major version to 10 very soon now anyway, we could decide to not support older versions and not bother to write this code.  For that case if we could have an error returned on the user size being too small that would be good to give an error up front.

Reply:
Noted. See request above https://github.com/DynamoRIO/dynamorio/pull/5835#discussion_r1289065632